Make GraphTask::owner_ atomic to avoid data races.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58457 Make c10::ThreadPool::available_ atomic.
* **#58449 Make GraphTask::owner_ atomic to avoid data races.**

Several TSAN tests were failing for distributed since `owner_` was not
atomic and was being accessed by several threads. As an example:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333.

Differential Revision: [D28496878](https://our.internmc.facebook.com/intern/diff/D28496878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28496878/)!

pritamdamania committed:
Make GraphTask::owner_ atomic to avoid data races.

Several TSAN tests were failing for distributed since `owner_` was not
atomic and was being accessed by several threads. As an example:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333.

Differential Revision: [D28496878](https://our.internmc.facebook.com/intern/diff/D28496878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28496878/)!

[ghstack-poisoned]

pritamdamania committed:
Update on "Make GraphTask::owner_ atomic to avoid data races."


Several TSAN tests were failing for distributed since `owner_` was not
atomic and was being accessed by several threads. As an example:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333.

Differential Revision: [D28496878](https://our.internmc.facebook.com/intern/diff/D28496878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28496878/)!

[ghstack-poisoned]

ezyang commented:
relevant issue #58442

ezyang reviewed (changes requested):
No, this looks wrong.

You can see that torch/csrc/autograd/engine.cpp is written very clearly to prevent synchronization on owner_. This is because owner_ is initialized at the construction of the first node task referring to the graph task, BEFORE we put it on the queue. I believe the invariant is that you aren't supposed to call execute_with_graph_task concurrently on the same graph task. 

In your quoted line https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333 you push the node task onto the queue, and then AFTERWARDS set the `owner_` on it.  That's seems wrong. Hopefully you also have the invariant that this function is called from a single thread?

pritamdamania committed:
Update on "Make GraphTask::owner_ atomic to avoid data races."


Several TSAN tests were failing for distributed since `owner_` was not
atomic and was being accessed by several threads. As an example:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333.

Differential Revision: [D28496878](https://our.internmc.facebook.com/intern/diff/D28496878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D28496878/)!

[ghstack-poisoned]

pritamdamania87 commented:
> In your quoted line https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L333 you push the node task onto the queue, and then AFTERWARDS set the owner_ on it. That's seems wrong. Hopefully you also have the invariant that this function is called from a single thread?

This invariant is not true in the distributed engine. This is part of multithreaded autograd support that we needed in the distributed engine where multiple RPCs might arrive on the same host and execute different parts of the autograd graph belonging to the same graph task. This invocation of the function can be called over RPC and be executed by different threads: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/autograd/engine/dist_engine.cpp#L519.

pritamdamania87 commented:
Closing in favor of https://github.com/pytorch/pytorch/pull/58625.
